### PR TITLE
Fix return handling of functions and events

### LINKED ISSUE #10: Handle async: 'parameterName'
runtime.sendMessage has
```
{
        "name": "sendMessage",
        "type": "function",
        "allowAmbiguousOptionalArguments": true,
        "allowedContexts": ["content", "devtools", "proxy"],
        "description": "[...]",
        "async": "responseCallback",`
[...]
```
current this function just returns `void` according to the definition. This is obviously not true.

### LINKED ISSUE #11: browser.events.Event callback parameters missing
https://dxr.mozilla.org/mozilla-central/source/toolkit/components/extensions/schemas/events.json#44

example method (there are more):

```json
{
  "name": "addListener",
  "type": "function",
  "description": "Registers an event listener <em>callback</em> to an event.",
  "parameters": [
    {
      "name": "callback",
      "type": "function",
      "description": "Called when an event occurs. The parameters of this function depend on the type of event."
    }
  ]
}
```

since the parameter is named `callback` it's ignored but maybe it shouldn't be if the function has no `async`?

this only really affects `browser.runtime.Port` since other things don't use this event definition

### PR DESCRIPTION
**Pull request by bomjacob**
Fixes #10 
Fixes #11 

- Allows other `async:` functions than just `async: 'callback'`
- Remove the callback parameter when it's gotten converted to a promise (this means we don't need to exclude it in the comment handler)
- And can therefore remove the callback exclude from the comment code

Also fixes (cause I stumbled upon the issues while fixes the stuff above)
- Custom events were ignoring custom callback return types
- Some webRequest events' callbacks can return either a BlockingResponse object or a promise resolving to such - this was not expressed in the schema at all as far as I could tell

### CODE REVIEW

**Comment by bomjacob** [comment]
Not 100% sure I merged that right... lemme check xD

File: index.js [hunk 1]
```diff
@@ -79,6 +79,19 @@ for (let path of [
...
+for (let path of [
+    ['webRequest', 'events', 'onAuthRequired'],
+    ['webRequest', 'events', 'onBeforeRequest'],
+    ['webRequest', 'events', 'onBeforeSendHeaders'],
+    ['webRequest', 'events', 'onHeadersReceived'],
+]) converter.edit(...path, x => {
+    // Return type of the callback is weirder than the schemas can express
+    x.returns.converterTypeOverride = 'BlockingResponse | Promise<BlockingResponse>';
+    // It's also optional, since you can choose to just listen to the event
+    x.returns.options = true;
```
Comment (epicgirl1998) [inline index.js]: is `options` a typo?

Comment (bomjacob) [reply index.js]: Yes indeed it is, whoops.

**Comment by bomjacob** [comment]
Hmm, in terms of some functions still having a `callback` it seems as if that only happens with unsupported functions - and is therefore probably not an issue.
``` typescript
/**
 * Returns currently registered rules.
 * @param eventName Name of the event this function affects.
 * @param webViewInstanceId If provided, this is an integer that uniquely identfies the <webview> associated with this function call.
 * @param [ruleIdentifiers] If an array is passed, only rules with identifiers contained in this array are returned.
 * @param callback Called with registered rules.
 * @deprecated Unsupported on Firefox at this time.
 */
getRules?(eventName: string, webViewInstanceId: number, ruleIdentifiers?: string[], callback: (rules: Rule[]) => void): void;
```
^ however, is an issue. Any clue why there's a required param after an optional one here?

**Comment by epicgirl1998** [comment]
- looks like `getRules` has no `async` key in the schema, and `ruleIdentifiers` is optional but the callback isn't

  i think even unsupported things should be converted as well as possible, since it's usable in other browsers with [`webextension-polyfill`](https://github.com/mozilla/webextension-polyfill)

  maybe `async` should be only used to get the callback name when it's present, but it should default to `'callback'` when missing, and fixes in `index.js` should be added for functions that actually take callbacks and don't return promises? there hopefully aren't many of them

- you can ignore what i said about `"async": true`, i thought it'd default to `"callback"` but none of the functions with `"async": true` have a callback

  for example: `browser.find.find`, the [mdn page](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/find/find) describes the return value but it's not there in [the schema](https://dxr.mozilla.org/mozilla-central/source/browser/components/extensions/schemas/find.json#26), so maybe the schema is outdated or firefox implements it in some different way?

  i searched for key names from the result object in the schema directories but there are no matches

i made [a script](https://gist.github.com/epicgirl1998/3640d019a59526a6ea90c74d06e3fab1) for searching for objects in the schema, maybe you'll find it useful (i mainly just use a simple text search though, unless it's something more complex)

**Comment by bomjacob** [comment]
Yeah i agree that they should be converted - and they are.
Question is if we should spent time figuring out if we need to add `async: 'callback'` outselves.

Yeah it seems like all the methods that firefox supports that chrome do not, do not define their Promises' return types in the schemas. I'm thinking we just do something like:
```
for (let func of [
    ['sessions', 'functions', 'setTabValue', 'Promise<void>'],
    ['sessions', 'functions', 'getTabValue', 'Promise<string | object | undefined>'],
    ['sessions', 'functions', 'removeTabValue', 'Promise<void>'],
    ['sessions', 'functions', 'setWindowValue', 'Promise<void>'],
    ['sessions', 'functions', 'getWindowValue', 'Promise<string | object | undefined>'],
    ['sessions', 'functions', 'removeWindowValue', 'Promise<void>'],
    ['sessions', 'functions', 'forgetClosedTab', 'Promise<void>'],
    ['sessions', 'functions', 'forgetClosedWindow', 'Promise<void>'],
    ['sessions', 'functions', 'getRecentlyClosed', 'Promise<Session[]>'],
    ['sessions', 'functions', 'restore', 'Promise<Session>'],
    ['sidebarAction', 'functions', 'close', 'Promise<void>'],
    ['sidebarAction', 'functions', 'open', 'Promise<void>'],
    ['sidebarAction', 'functions', 'setPanel', 'Promise<void>'],
    ['sidebarAction', 'functions', 'setIcon', 'Promise<void>'],
    ['sidebarAction', 'functions', 'setTitle', 'Promise<void>'],
    ['sidebarAction', 'functions', 'getPanel', 'Promise<string>'],
    ['sidebarAction', 'functions', 'getTitle', 'Promise<string>'],
    [... continued]
]) converter.edit(func[0], func[1], func[2], y => {
    y.returns = {converterTypeOverride: func[3]};
    return y;
});
```

**Pull request merged**